File: js/IntcodeComputer.js
module.exports = {
  program(registries) {
    return new Computer(program(registries, []));
  }
};

const RUNNING = "RUNNING";
const HALTED = "HALTED";
const WAITING_FOR_INPUT = "WAITING_FOR_INPUT";

class Computer {
  /** @param {ReturnType<typeof program>} program */
  constructor({ registries, input, output, ...rest }) {
    this._program = {
      registries: registries.slice(),
      input: input.slice(),
      output: output.slice(),
      ...rest
    };
  }
  isHalted() {
    return this._program.status === HALTED;
  }
  /** @param {number[]} input */
  withInput(input) {
    return new Computer({ ...this._program, input });
  }
  /** @param {number[]} input */
  pushInput(input) {
    this._program.input.push(...input);
    return this;
  }
  run() {
    return runProgram(this._program);
  }
}

/**
 * @param {number[]} registries
 * @param {number[]} input
 */
function program(registries, input) {
  return {
    registries,
    input,
    output: [],
    position: 0,
    status: RUNNING,
    relativeBase: 0
  };
}

const SUM = "SUM";
const MUL = "MUL";
const INPUT = "INPUT";
const OUTPUT = "OUTPUT";
const HALT = "HALT";
const JIT = "JIT";
const JIF = "JIF";
const LT = "LT";
const EQ = "EQ";
const ADJUST_RB = "ADJUST_RB";

const sum = (left, right, position) => ({ type: SUM, left, right, position });
const mul = (left, right, position) => ({ type: MUL, left, right, position });
const input = position => ({ type: INPUT, position });
const output = left => ({ type: OUTPUT, left });
const jit = (left, right) => ({ type: JIT, left, right });
const jif = (left, right) => ({ type: JIF, left, right });
const lt = (left, right, position) => ({ type: LT, left, right, position });
const eq = (left, right, position) => ({ type: EQ, left, right, position });
const adjustRb = left => ({ type: ADJUST_RB, left });
const halt = () => ({ type: HALT });

/** @returns {ReturnType<typeof program>} */
function runProgram(program) {
  if (program.status === WAITING_FOR_INPUT && program.input.length > 0) {
    program.status = RUNNING;
  }
  while (program.status === RUNNING) {
    runCommand(program, parseCommand(program));
  }
  return program;
}

function runCommand(program, { type, left, right, position }) {
  switch (type) {
    case SUM:
      program.registries[position] = left + right;
      program.position += 4;
      break;
    case MUL:
      program.registries[position] = left * right;
      program.position += 4;
      break;
    case INPUT:
      if (program.input.length === 0) {
        program.status = WAITING_FOR_INPUT;
      } else {
        program.registries[position] = program.input.shift();
        program.position += 2;
      }
      break;
    case OUTPUT:
      program.output.push(left);
      program.position += 2;
      break;
    case JIT:
      program.position = left ? right : program.position + 3;
      break;
    case JIF:
      program.position = left ? program.position + 3 : right;
      break;
    case LT:
      program.registries[position] = Number(left < right);
      program.position += 4;
      break;
    case EQ:
      program.registries[position] = Number(left === right);
      program.position += 4;
      break;
    case ADJUST_RB:
      program.relativeBase += left;
      program.position += 2;
      break;
    case HALT:
      program.status = HALTED;
      return;
  }
}

function parseCommand({ registries, position, relativeBase }) {
  const registry = registries[position];
  const command = registry % 100;
  const m1 = Math.floor(registry / 100) % 10;
  const m2 = Math.floor(registry / 1000) % 10;
  const m3 = Math.floor(registry / 10000) % 10;
  const p1 = () => param(registries, m1, position + 1, relativeBase);
  const p2 = () => param(registries, m2, position + 2, relativeBase);
  const p3 = () => param(registries, m3, position + 3, relativeBase);
  switch (command) {
    case 1:
      return sum(registries[p1()], registries[p2()], p3());
    case 2:
      return mul(registries[p1()], registries[p2()], p3());
    case 3:
      return input(p1());
    case 4:
      return output(registries[p1()]);
    case 5:
      return jit(registries[p1()], registries[p2()]);
    case 6:
      return jif(registries[p1()], registries[p2()]);
    case 7:
      return lt(registries[p1()], registries[p2()], p3());
    case 8:
      return eq(registries[p1()], registries[p2()], p3());
    case 9:
      return adjustRb(registries[p1()]);
    case 99:
      return halt();
    default:
      throw new TypeError("Unknown command: " + command);
  }
}

function param(registries, mode, position, relativeBase) {
  if (mode === 1) {
    return position;
  }
  if (mode === 2) {
    return relativeBase + registries[position];
  }
  return registries[position];
}

File: haskell/src/IntcodeComputer.hs
module IntcodeComputer
    ( Program(..)
    , ProgramStatus(..)
    , Int
    , runProgramm
    , fromRegistries
    )
where

import           Lib
import           Data.Map                      as M
import           Debug.Trace
import           Data.Maybe

data ProgramStatus = Running | Halted | WaitingForInput deriving (Show, Eq)

type Regs = M.Map Int Int

data Program = Program {
    registries :: Regs,
    input :: [Int],
    output :: [Int],
    position :: Int,
    status :: ProgramStatus,
    relativeBase :: Int
} deriving Show

data Command =
    SumC Int Int Int
    | ProdC Int Int Int
    | InputC Int
    | OutputC Int
    | JumpIfTrueC Int Int
    | JumpIfFalseC Int Int
    | LessThanC Int Int Int
    | EqualsC Int Int Int
    | AdjustRB Int
    | Halt
    deriving Show

fromRegistries :: [Int] -> Program
fromRegistries registries = Program { registries   = regs
                                    , input        = []
                                    , output       = []
                                    , position     = 0
                                    , status       = Running
                                    , relativeBase = 0
                                    }
    where regs = M.fromList $ zip [0 ..] registries

runProgramm :: Program -> Program
runProgramm program =
    let command  = parseCommand program
        program' = runCommand program command
    in  case status program' of
            Halted          -> program'
            WaitingForInput -> program'
            Running         -> runProgramm program'

runCommand :: Program -> Command -> Program
runCommand program@Program { registries = regs, position = pos } command =
    case command of
        (SumC l r p) -> program { registries = M.insert p (l + r) regs
                                , position   = pos + 4
                                , status     = Running
                                }
        (ProdC l r p) -> program { registries = M.insert p (l * r) regs
                                 , position   = pos + 4
                                 , status     = Running
                                 }
        (InputC p) -> case input program of
            [] -> program { status = WaitingForInput }
            _  -> program { registries = M.insert p (head $ input program) regs
                          , input      = tail $ input program
                          , position   = pos + 2
                          , status     = Running
                          }
        (OutputC p) -> program { output   = output program ++ [p]
                               , position = pos + 2
                               , status   = Running
                               }
        (JumpIfTrueC l p) -> program { position = if l == 0 then pos + 3 else p
                                     , status   = Running
                                     }
        (JumpIfFalseC l p) -> program
            { position = if l == 0 then p else pos + 3
            , status   = Running
            }
        (LessThanC l r p) -> program
            { registries = M.insert p (if l < r then 1 else 0) regs
            , position   = pos + 4
            , status     = Running
            }
        (EqualsC l r p) -> program
            { registries = M.insert p (if l == r then 1 else 0) regs
            , position   = pos + 4
            , status     = Running
            }
        (AdjustRB p) -> program { relativeBase = relativeBase program + p
                                , status       = Running
                                , position     = pos + 2
                                }
        Halt -> program { status = Halted }

getP :: Int -> Regs -> Int
getP pos regs = fromMaybe 0 (M.lookup pos regs)

parseCommand :: Program -> Command
parseCommand p@Program { registries = regs, position = pos } = case command of
    1  -> SumC (get p1) (get p2) p3
    2  -> ProdC (get p1) (get p2) p3
    3  -> InputC p1
    4  -> OutputC (get p1)
    5  -> JumpIfTrueC (get p1) (get p2)
    6  -> JumpIfFalseC (get p1) (get p2)
    7  -> LessThanC (get p1) (get p2) p3
    8  -> EqualsC (get p1) (get p2) p3
    9  -> AdjustRB (get p1)
    99 -> Halt
    _  -> error $ "Unknown command " ++ show command
  where
    get             = flip getP regs
    commandAndModes = get pos
    command         = commandAndModes `mod` 100
    modes           = commandAndModes `div` 100
    p1              = param p (modes `mod` 10) (pos + 1)
    p2              = param p ((modes `div` 10) `mod` 10) (pos + 2)
    p3              = param p ((modes `div` 100) `mod` 10) (pos + 3)

param :: Program -> Int -> Int -> Int
param Program { registries = regs, relativeBase = rb } mode pos = case mode of
    0 -> getP pos regs
    1 -> pos
    2 -> rb + getP pos regs
